File: apps/api/src/services/manual_entry.ts
import crypto from 'crypto';
import { SupabaseClient } from '@supabase/supabase-js';

export interface ManualLineItemInput {
  description: string;
  charge_code?: string;
  amount: number;
}

export interface ManualAccessorialInput {
  type: string;
  amount: number;
  authorized?: boolean;
}

export interface ManualInvoiceInput {
  customerId: string;
  carrier: string;
  pro_number: string;
  invoice_number: string;
  invoice_date: string; // YYYY-MM-DD
  origin_zip: string;
  dest_zip: string;
  billed_weight: number;
  billed_class?: number;
  line_items: ManualLineItemInput[];
  accessorials?: ManualAccessorialInput[];
  fsc_amount?: number;
  fsc_pct?: number;
  invoice_total: number;
  notes?: string;
}

export interface ManualEntryResult {
  success: boolean;
  invoiceId?: string;
  error?: string;
  isDuplicate?: boolean;
}

/**
 * Validates and stores a manually entered freight invoice.
 * Enters the same pipeline state ('pending') as an emailed/uploaded invoice.
 */
export async function createManualInvoice(
  input: ManualInvoiceInput,
  supabase: SupabaseClient
): Promise<ManualEntryResult> {
  // 1. Basic validation
  if (!input.customerId || !input.carrier || !input.pro_number || !input.invoice_number) {
    return { success: false, error: 'Missing required fields (carrier, pro_number, invoice_number)' };
  }

  if (input.invoice_total <= 0) {
    return { success: false, error: 'Invoice total must be greater than zero' };
  }

  // 2. Dedup constraint check against DB unique index (customer_id, carrier, invoice_number, pro_number)
  const { data: existing } = await supabase
    .from('invoices')
    .select('id')
    .eq('customer_id', input.customerId)
    .eq('carrier', input.carrier)
    .eq('invoice_number', input.invoice_number)
    .eq('pro_number', input.pro_number)
    .maybeSingle();

  if (existing) {
    return { 
      success: false, 
      isDuplicate: true, 
      error: `Invoice ${input.invoice_number} (PRO ${input.pro_number}) already exists for this carrier` 
    };
  }

  // 3. Construct canonical InvoiceJSON structure
  const invoiceId = crypto.randomUUID();
  const canonicalInvoiceJson = {
    carrier: input.carrier,
    pro_number: input.pro_number,
    invoice_number: input.invoice_number,
    invoice_date: input.invoice_date,
    origin_zip: input.origin_zip,
    dest_zip: input.dest_zip,
    billed_weight: input.billed_weight,
    billed_class: input.billed_class,
    line_items: input.line_items || [],
    accessorials: input.accessorials || [],
    fsc_amount: input.fsc_amount,
    fsc_pct: input.fsc_pct,
    invoice_total: input.invoice_total,
    notes: input.notes,
    raw_text_hash: crypto.createHash('sha256').update(JSON.stringify(input)).digest('hex'),
    manual_entry: true
  };

  // 4. Insert into invoices table
  const { data: inserted, error } = await supabase
    .from('invoices')
    .insert({
      id: invoiceId,
      customer_id: input.customerId,
      source: 'manual',
      carrier: input.carrier,
      pro_number: input.pro_number,
      invoice_number: input.invoice_number,
      invoice_date: input.invoice_date,
      invoice_total: input.invoice_total,
      parsed_json: canonicalInvoiceJson,
      parse_confidence: 1.0, // Human entered and confirmed
      file_path: null, // No source PDF for pure manual stragglers/faxes
      status: 'pending' // Ready for audit engine execution
    })
    .select('id')
    .single();

  if (error) {
    return { success: false, error: error.message };
  }

  return {
    success: true,
    invoiceId: inserted.id
  };
}

File: apps/worker/manual_entry.py
"""
Manual invoice entry service for Python runtime.
Guarantees manual invoices enter the exact same pipeline state ('pending') as emailed/uploaded invoices.
"""
import hashlib
import json
import uuid
from typing import Any, Dict, Optional, Tuple


def build_canonical_manual_invoice(data: Dict[str, Any]) -> Dict[str, Any]:
    """
    Constructs a validated canonical InvoiceJSON structure from manual form input.
    """
    required = ["carrier", "pro_number", "invoice_number", "invoice_date", "invoice_total"]
    for field in required:
        if not data.get(field):
            raise ValueError(f"Missing required manual invoice field: {field}")

    if float(data["invoice_total"]) <= 0:
        raise ValueError("Invoice total must be greater than zero")

    raw_hash = hashlib.sha256(json.dumps(data, sort_keys=True).encode("utf-8")).hexdigest()

    return {
        "carrier": data["carrier"].strip(),
        "pro_number": data["pro_number"].strip(),
        "invoice_number": data["invoice_number"].strip(),
        "invoice_date": data["invoice_date"].strip()[:10],
        "origin_zip": data.get("origin_zip", "").strip(),
        "dest_zip": data.get("dest_zip", "").strip(),
        "billed_weight": float(data.get("billed_weight", 0.0)),
        "billed_class": float(data["billed_class"]) if data.get("billed_class") is not None else None,
        "line_items": data.get("line_items", []),
        "accessorials": data.get("accessorials", []),
        "fsc_amount": float(data["fsc_amount"]) if data.get("fsc_amount") is not None else None,
        "fsc_pct": float(data["fsc_pct"]) if data.get("fsc_pct") is not None else None,
        "invoice_total": float(data["invoice_total"]),
        "notes": data.get("notes"),
        "raw_text_hash": raw_hash,
        "manual_entry": True
    }


def insert_manual_invoice(conn, customer_id: str, data: Dict[str, Any]) -> Tuple[bool, Optional[str], Optional[str]]:
    """
    Inserts manual invoice into Supabase Postgres database.
    Returns: (success, invoice_id, error_message)
    """
    try:
        canonical_json = build_canonical_manual_invoice(data)
    except Exception as val_err:
        return False, None, str(val_err)

    invoice_id = str(uuid.uuid4())
    cur = conn.cursor()

    # Check duplicate constraint
    cur.execute("""
        SELECT id FROM invoices
        WHERE customer_id = %s AND carrier = %s AND invoice_number = %s AND pro_number = %s;
    """, (customer_id, canonical_json["carrier"], canonical_json["invoice_number"], canonical_json["pro_number"]))

    if cur.fetchone():
        return False, None, f"Duplicate invoice: {canonical_json['invoice_number']} already exists"

    try:
        cur.execute("""
            INSERT INTO invoices (
                id, customer_id, source, carrier, pro_number, invoice_number,
                invoice_date, invoice_total, parsed_json, parse_confidence,
                file_path, status
            ) VALUES (
                %s, %s, 'manual', %s, %s, %s,
                %s, %s, %s, 1.000,
                NULL, 'pending'
            ) RETURNING id;
        """, (
            invoice_id,
            customer_id,
            canonical_json["carrier"],
            canonical_json["pro_number"],
            canonical_json["invoice_number"],
            canonical_json["invoice_date"],
            canonical_json["invoice_total"],
            json.dumps(canonical_json)
        ))
        conn.commit()
        return True, invoice_id, None
    except Exception as e:
        conn.rollback()
        return False, None, str(e)
